File: frontend/src/services/api/knowledge-graph.ts
/**
 * Knowledge graph API module.
 * Provides entity and relation management for agent knowledge graphs.
 */
import { apiFetch } from './client';

export interface KGEntity {
  id: string;
  agent_id: string;
  name: string;
  entity_type: string;
  properties: Record<string, unknown>;
  mention_count: number;
  importance_score: number;
  first_seen: string;
  last_seen: string;
  relations?: KGRelation[];
}

export interface KGRelation {
  id: string;
  agent_id: string;
  source_id: string;
  target_id: string;
  relation_type: string;
  properties: Record<string, unknown>;
  confidence: number;
  mention_count: number;
  source_name?: string;
  target_name?: string;
}

export interface KGSubgraph {
  entities: KGEntity[];
  relations: KGRelation[];
}

export interface KGStats {
  entity_count: number;
  relation_count: number;
  top_entities: KGEntity[];
  last_consolidation: string | null;
}

export const knowledgeGraphApi = {
  listEntities: (agentId: string) =>
    apiFetch<{ entities: KGEntity[]; total: number }>(
      `/admin/agents/${agentId}/knowledge/entities`
    ),

  getEntity: (agentId: string, entityId: string) =>
    apiFetch<KGEntity>(`/admin/agents/${agentId}/knowledge/entities/${entityId}`),

  deleteEntity: (agentId: string, entityId: string) =>
    apiFetch<{ message: string }>(
      `/admin/agents/${agentId}/knowledge/entities/${entityId}`,
      { method: 'DELETE' }
    ),

  queryGraph: (agentId: string, seed: string, hops = 1) => {
    const params = new URLSearchParams({ seed, hops: String(hops) });
    return apiFetch<KGSubgraph>(`/admin/agents/${agentId}/knowledge/graph?${params}`);
  },

  searchEntities: (agentId: string, query: string) => {
    const params = new URLSearchParams({ q: query });
    return apiFetch<{ entities: KGEntity[]; total: number }>(
      `/admin/agents/${agentId}/knowledge/search?${params}`
    );
  },

  getStats: (agentId: string) =>
    apiFetch<KGStats>(`/admin/agents/${agentId}/knowledge/stats`),

  triggerConsolidation: (agentId: string) =>
    apiFetch<{ consolidated: number; details: Array<Record<string, unknown>> }>(
      `/admin/agents/${agentId}/knowledge/consolidate`,
      { method: 'POST' }
    ),
};

File: backend/app/models/knowledge_graph.py
"""Pydantic models for knowledge graph API."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class KGEntityType(str, Enum):
    PERSON = "person"
    CONCEPT = "concept"
    TOOL = "tool"
    PROJECT = "project"
    ORGANIZATION = "organization"
    TECHNOLOGY = "technology"
    OTHER = "other"


class KGEntity(BaseModel):
    id: str
    agent_id: str
    name: str
    entity_type: str
    properties: dict = Field(default_factory=dict)
    mention_count: int = 1
    importance_score: float = 0.5
    first_seen: Optional[str] = None
    last_seen: Optional[str] = None


class KGRelation(BaseModel):
    id: str
    agent_id: str
    source_id: str
    target_id: str
    relation_type: str
    properties: dict = Field(default_factory=dict)
    confidence: float = 0.5
    mention_count: int = 1


class KGSubgraph(BaseModel):
    entities: list[KGEntity] = Field(default_factory=list)
    relations: list[KGRelation] = Field(default_factory=list)


class KGAgentPath(BaseModel):
    agent_id: str = Field(..., description="Agent ID")


class KGEntityPath(BaseModel):
    agent_id: str = Field(..., description="Agent ID")
    entity_id: str = Field(..., description="KG entity ID")


class KGGraphQuery(BaseModel):
    seed: str = Field(..., description="Seed entity name for graph traversal")
    hops: int = Field(1, ge=1, le=3, description="Traversal depth")


class KGStatsResponse(BaseModel):
    entity_count: int = 0
    relation_count: int = 0
    top_entities: list[dict] = Field(default_factory=list)
    last_extraction: Optional[str] = None
